docs/proposals/20210325-imagepredownload-during-inplace-update.md

```diff
@@ -0,0 +1,113 @@
+
+If CloneSet is going to update Pods in-place,
+the controller could pre-download the new image on the Nodes of all Pods in the CloneSet.
+
+## Table of Contents
+
+A table of contents is helpful for quickly jumping to sections of a proposal and for highlighting
+any additional information provided beyond the standard proposal template.
+[Tools for generating](https://github.com/ekalinin/github-markdown-toc) a table of contents from markdown are available.
+
+- [Pre-download image during in-place update](#pre-download-image-during-in-place-update)
+  - [Table of Contents](#table-of-contents)
+  - [Motivation](#motivation)
+  - [Proposal](#proposal)
+    - [User Stories](#user-stories)
+      - [Story 1](#story-1)
+    - [Requirements](#requirements)
+    - [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+    - [Risks and Mitigations](#risks-and-mitigations)
+  - [Implementation History](#implementation-history)
+
+## Motivation
+
+If the image has been pre-downloaded on Nodes, we can significantly accelerate the speed of in-place update,
+for Kubelet has not to pull the image again.
+
+## Proposal
+
+Add a new feature-gate: `PreDownloadImageForInPlaceUpdate`.
+
+If it is opened, CloneSet will create a `ImagePullJob` when it starts a new in-place update:
+
+```yaml
+apiVersion: apps.kruise.io/v1alpha1
+kind: ImagePullJob
+metadata:
+  namespace: cloneset-namespace
+  name: cloneset-name-{random-id}
+spec:
+  image: nginx:1.9.1
+  parallelism: 1
+  podSelector:
+    matchLabels:
+      {the selector of CloneSet}
+    matchExpressions:
+      - key: controller-revision-hash
+        operator: NotIn
+        values:
+        - {the updateRevision of CloneSet}
+  completionPolicy:
+    type: Always
+  pullPolicy:
+    backoffLimit: 1
+    timeoutSeconds: 300
+```
+
+Default values:
+
+- The requested parallelism `parallelism` defaults to `1`.
+- The timeout for pulling image `timeoutSeconds` defaults to `300`.
+
+Also, users can configure them in their CloneSet:
+
+```yaml
+apiVersion: apps.kruise.io/v1alpha1
+kind: CloneSet
+metadata:
+  annotations:
+    cloneset.apps.kruise.io/image-predownload-parallelism: 10
+    cloneset.apps.kruise.io/image-predownload-timeout-seconds: 600
+```
+
+### User Stories
+
+#### Story 1
+
+Image pulling is the most time-consuming stage in an in-place update progress.
+
+If the new image has already been downloaded on the Node, it makes in-place update more efficient.
+
+### Requirements
+
+It relies on `ImagePullJob` and `NodeImage` to pre-download images.
+
+So if the `KruiseDaemon` feature-gate is closed, `PreDownloadImageForInPlaceUpdate` can neither be opened.
+
+### Implementation Details/Notes/Constraints
+
+1. User updates an image in CloneSet template.
+2. CloneSet controller creates an ImagePullJob to pull the new image on the Nodes of its Pods.
+3. When sorting Pods to update, CloneSet controller will firstly sort them by the image has already downloaded on their Nodes.
```

Comment:
since there is many sorting criteria of pods to update, plz give the relative priority of each sorting criteria. And how do we ensure the sorting logic of  imagepulljob choosing the next node to puliing the image, are the same with the sorting logic of choose the next pod to update. if two sorting are not in sync, the predownload be less efficient